phase1/design_document.md

```diff
@@ -0,0 +1,102 @@
+    - For example, when presenting a defender, we aren't interested in their goalkeeping abilities.
+  - To avoid multiple switch statements in our program, we use polymorphism instead, thus avoiding a [code smell](https://refactoring.guru/smells/switch-statements).
+    - The `PlayerPresenter` method is overloaded to allow different outputs depending on the subclass of `Player` that we pass in.
+- Implemented the **builder design pattern** for the different types of searches
+  - *Details/justification...*
+- Replaced `Scout` with a `User` class
+  - We had always thought of `Scout`s as the users of the program, so this class name didn't accurately represent what we wanted the class to do.
+    - *i.e.* It doesn't make much sense for a `Scout` class to have a username and password.
+- Made our Business Data (entities) serializable
+  - *Details/justification...*
+- Reworked our search/input classes
+  - *Details/justification...*
+
+## Clean Architecture
+
+*TODO: Draw a chart of our classes in the layers of clean architecture w/ dependency arrows*
+
+*TODO: Brief explanation of chart*
+
+## SOLID Design
+
+- Single Responsibility Principle
+  - *Explanation of how our project follows this...*
+  - *Maybe mention how we can't follow this sometimes; if our dataset changes, we might have to change every single part of our program*
+- Open/Closed Principle
+  - *Explanation of how our project follows this...*
+- Liskov Substitution Principle
+  - *Explanation of how our project follows this...*
+- Interface Segregation Principle
+  - *Explanation of how our project follows this...*
+- Dependency Inversion
+  - *Explanation of how our project follows this...*
+
+## Packaging Strategy
+
+*TODO: Why did we go with package by component? (might help to explain by process of elimination)*
+
+## Design Patterns
+
+In the previous phase, we implemented the adapter design pattern. *Explanation...*
+
+In this phase, we implemented the factory design pattern and builder design pattern (see [above](#major-design-decisions) for details/explanations). 
+Below is an outline of the role that each class plays in satisfying the design pattern:
+
+- factory design pattern:
+  - factory class = `PlayerFactory`
+  - common base class = `Player`
+  - products = `Forward`, `Midfielder`, `Defender`, `Goalkeeper` (subclasses of `Player`)
+- builder design pattern:
+  - *...*
+
+We also considered implementing the decorator design pattern in this phase.
+The idea was to have a `BasePlayerStatsCalculator` class which can calculate the overall rating of a `Player` or `Team`, taking into account **all** skill attributes.
+Then, we would have several decorator classes like `OffensiveRatingDecorator` and `GoalkeepingRatingDecorator` that would provide the additional functionality of generating a rating based on a **subset** of skill attributes.
+*However, we decided against this, since...*
```

Comment:
@m-parvaneh Can you add a little explanation for why the decorator design pattern wasn't applicable to our project?

Reply:
Of course, I will add an explanation as to why I made that decision.